File: src/bench_patchbay.zig
//! In-process microbench for patchbay rule evaluation.
//!
//! Measures pure eval cost (no router, no TCP, no parser, no syscalls) so
//! a change to the evaluator or builtins shows up cleanly. The publisher
//! is a no-op that just bumps a counter, so each iteration is bounded by
//! `RuleSet.run` + the body's ops.
//!
//! Run with: `zig build bench-patchbay -- [MODE] [N] [PUBS]`
//!   MODE  one of:
//!         pass   - body is just (publish! subject payload). Floor: dispatch + emit.
//!         gate   - (-> payload-float (squelch) (publish! ...)). Per-rule state hashmap.
//!         window - (-> payload-float (moving-avg 10) (publish! ...)). Ring updates.
//!         json   - (json-demux! "a" "b" "c" payload) over a small object.
//!         mixed  - loads ./patchbay.edn from cwd and runs that.
//!   N     rule count (default 1; ignored in `mixed` since the file fixes it)
//!   PUBS  publishes per run (default 1_000_000)
//!
//! Subject is `bench.k<i % 1024>` so literal-index lookups are cold-ish
//! and wildcard rules ("sensors.*", ">") still match. Payload is a short
//! numeric string for the float-coerce modes, a tiny JSON object for json.

const std = @import("std");
const Io = std.Io;
const patchbay = @import("patchbay");

const eval = patchbay.eval;
const Rule = eval.Rule;
const RuleSet = eval.RuleSet;
const Context = eval.Context;
const Publisher = eval.Publisher;

const Mode = enum { pass, gate, window, json, mixed };

/// Monotonic ns. `std.time.nanoTimestamp` was removed in Zig 0.16.
fn monoNs() i128 {
    var ts: std.posix.timespec = undefined;
    _ = std.posix.system.clock_gettime(std.posix.CLOCK.MONOTONIC, &ts);
    return @as(i128, ts.sec) * std.time.ns_per_s + @as(i128, ts.nsec);
}

const NoopPublisher = struct {
    count: u64 = 0,

    fn publish(ctx: *anyopaque, _: []const u8, _: []const u8) anyerror!void {
        const self: *NoopPublisher = @ptrCast(@alignCast(ctx));
        self.count += 1;
    }

    fn publisher(self: *NoopPublisher) Publisher {
        return .{ .ctx = self, .publish_fn = publish };
    }
};

fn parseMode(s: []const u8) ?Mode {
    if (std.mem.eql(u8, s, "pass")) return .pass;
    if (std.mem.eql(u8, s, "gate")) return .gate;
    if (std.mem.eql(u8, s, "window")) return .window;
    if (std.mem.eql(u8, s, "json")) return .json;
    if (std.mem.eql(u8, s, "mixed")) return .mixed;
    return null;
}

/// Build a synthetic patchbay source for `mode` with `n` rules. All rules
/// share the filter "sensors.*" so the timed loop's "sensors.k<i>" subject
/// matches every one (worst-case: every rule fires per PUB).
fn buildSource(arena: std.mem.Allocator, mode: Mode, n: usize) ![]const u8 {
    var buf: std.ArrayList(u8) = .empty;
    var i: usize = 0;
    while (i < n) : (i += 1) {
        switch (mode) {
            .pass => try buf.print(
                arena,
                "(on \"sensors.*\" (publish! (subject-append \"r{d}\") payload))\n",
                .{i},
            ),
            .gate => try buf.print(
                arena,
                "(on \"sensors.*\" (-> payload-float (squelch) (publish! (subject-append \"r{d}\"))))\n",
                .{i},
            ),
            .window => try buf.print(
                arena,
                "(on \"sensors.*\" (-> payload-float (moving-avg 10) (publish! (subject-append \"r{d}\"))))\n",
                .{i},
            ),
            .json => try buf.print(
                arena,
                "(on \"sensors.*\" (json-demux! \"a\" \"b\" \"c\" payload))\n",
                .{},
            ),
            .mixed => unreachable,
        }
    }
    return buf.toOwnedSlice(arena);
}

pub fn main(init: std.process.Init) !void {
    const gpa = init.gpa;
    const fsio = init.io;

    var it = try init.minimal.args.iterateAllocator(gpa);
    defer it.deinit();
    _ = it.skip(); // exe name

    const mode_arg: []const u8 = it.next() orelse "pass";
    const mode = parseMode(mode_arg) orelse {
        std.debug.print("bad mode '{s}' (want pass|gate|window|json|mixed)\n", .{mode_arg});
        return error.BadArgs;
    };
    const n: usize = if (it.next()) |s| try std.fmt.parseInt(usize, s, 10) else 1;
    const pubs: usize = if (it.next()) |s| try std.fmt.parseInt(usize, s, 10) else 1_000_000;

    if (n == 0 and mode != .mixed) {
        std.debug.print("N must be > 0\n", .{});
        return error.BadArgs;
    }

    // Arena for the source string + parsed AST + RuleSet index. Lives the
    // length of the run; everything else (per-rule state, the per-message
    // arena) uses gpa.
    var src_arena_state: std.heap.ArenaAllocator = .init(gpa);
    defer src_arena_state.deinit();
    const src_arena = src_arena_state.allocator();

    const source: []const u8 = switch (mode) {
        .mixed => try readFile(fsio, src_arena, "patchbay.edn"),
        else => try buildSource(src_arena, mode, n),
    };

    const rules = try eval.loadRules(src_arena, source);
    defer eval.deinitRules(rules, gpa);

    const ruleset = try eval.buildRuleSet(src_arena, rules);

    var noop: NoopPublisher = .{};

    // Per-message arena reused across iterations; reset each PUB like the
    // server does. Allocations inside eval (subject-append, str-concat,
    // json-demux output) live here.
    var msg_arena_state: std.heap.ArenaAllocator = .init(gpa);
    defer msg_arena_state.deinit();

    // Subject + payload buffers reused across iterations.
    var subj_buf: [64]u8 = undefined;
    const float_payloads = [_][]const u8{
        "1.0", "1.1", "1.2", "1.3", "1.4", "1.5", "1.6", "1.7",
        "2.0", "2.1", "2.2", "2.3", "2.4", "2.5", "2.6", "2.7",
    };

    const pickPayload = struct {
        fn f(m: Mode, i: usize, floats: []const []const u8) []const u8 {
            return switch (m) {
                .json => "{\"a\":1,\"b\":2,\"c\":3}",
                else => floats[i % floats.len],
            };
        }
    }.f;

    // Warm up: stateful ops (squelch hashmap, moving-avg ring) measure
    // different work cold vs warm. 10k iterations is plenty to populate
    // every per-rule slot for the "sensors.k0..k1023" subject space.
    var warm_i: usize = 0;
    while (warm_i < 10_000) : (warm_i += 1) {
        _ = msg_arena_state.reset(.retain_capacity);
        const subj = try std.fmt.bufPrint(&subj_buf, "sensors.k{d}", .{warm_i % 1024});
        var ctx: Context = .{
            .subject = subj,
            .payload = pickPayload(mode, warm_i, &float_payloads),
            .publisher = noop.publisher(),
            .arena = msg_arena_state.allocator(),
            .gpa = gpa,
        };
        try ruleset.run(&ctx);
    }
    const warm_emits = noop.count;
    noop.count = 0;

    // Timed loop.
    const start = monoNs();
    var i: usize = 0;
    while (i < pubs) : (i += 1) {
        _ = msg_arena_state.reset(.retain_capacity);
        const subj = try std.fmt.bufPrint(&subj_buf, "sensors.k{d}", .{i % 1024});
        var ctx: Context = .{
            .subject = subj,
            .payload = pickPayload(mode, i, &float_payloads),
            .publisher = noop.publisher(),
            .arena = msg_arena_state.allocator(),
            .gpa = gpa,
        };
        try ruleset.run(&ctx);
    }
    const elapsed_ns = monoNs() - start;

    const elapsed_s: f64 = @as(f64, @floatFromInt(elapsed_ns)) / 1e9;
    const rate: f64 = @as(f64, @floatFromInt(pubs)) / elapsed_s;
    const ns_per_pub: f64 = @as(f64, @floatFromInt(elapsed_ns)) / @as(f64, @floatFromInt(pubs));
    const emits_per_pub: f64 = @as(f64, @floatFromInt(noop.count)) / @as(f64, @floatFromInt(pubs));

    std.debug.print(
        "mode={s} rules={d} | {d} pubs in {d:.3}s = {d:.0} pubs/s ({d:.1} ns/pub, {d:.2} emits/pub, warm_emits={d})\n",
        .{ @tagName(mode), rules.len, pubs, elapsed_s, rate, ns_per_pub, emits_per_pub, warm_emits },
    );
}

fn readFile(io: Io, arena: std.mem.Allocator, path: []const u8) ![]u8 {
    const cwd = Io.Dir.cwd();
    var file = try cwd.openFile(io, path, .{});
    defer file.close(io);
    const len = try file.length(io);
    const buf = try arena.alloc(u8, @intCast(len));
    var file_reader = file.reader(io, buf);
    try file_reader.interface.readSliceAll(buf);
    return buf;
}

File: src/bench_patchbay.c
#define _POSIX_C_SOURCE 200809L

#include "buf.h"
#include "fs.h"
#include "pb_program.h"

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef enum bench_mode {
    BENCH_PASS,
    BENCH_GATE,
    BENCH_WINDOW,
    BENCH_JSON,
    BENCH_MIXED,
} bench_mode;

typedef struct noop_publisher {
    uint64_t count;
} noop_publisher;

enum { BENCH_SCRATCH_RETAIN_BYTES = 4 * 1024 * 1024 };

static const char *mode_name(bench_mode mode) {
    switch (mode) {
    case BENCH_PASS: return "pass";
    case BENCH_GATE: return "gate";
    case BENCH_WINDOW: return "window";
    case BENCH_JSON: return "json";
    case BENCH_MIXED: return "mixed";
    }
    return "?";
}

static bool parse_mode(const char *s, bench_mode *out) {
    if (strcmp(s, "pass") == 0) *out = BENCH_PASS;
    else if (strcmp(s, "gate") == 0) *out = BENCH_GATE;
    else if (strcmp(s, "window") == 0) *out = BENCH_WINDOW;
    else if (strcmp(s, "json") == 0) *out = BENCH_JSON;
    else if (strcmp(s, "mixed") == 0) *out = BENCH_MIXED;
    else return false;
    return true;
}

static bool publish_noop(void *ctx, pb_slice subject, pb_slice payload) {
    (void)subject;
    (void)payload;
    noop_publisher *pub = ctx;
    pub->count += 1;
    return true;
}

static uint64_t mono_ns(void) {
    struct timespec ts = {0};
    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
        return 0;
    }
    return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

static bool append_line(mb_buf *buf, const char *fmt, size_t i) {
    char tmp[256];
    const int n = snprintf(tmp, sizeof tmp, fmt, i);
    if (n < 0 || (size_t)n >= sizeof tmp) {
        return false;
    }
    return mb_buf_append(buf, tmp, (size_t)n);
}

static bool build_source(bench_mode mode, size_t n, char **out, size_t *out_len) {
    mb_buf buf = {0};
    for (size_t i = 0; i < n; i += 1) {
        bool ok = false;
        switch (mode) {
        case BENCH_PASS:
            ok = append_line(&buf,
                             "(on \"sensors.*\" (publish! (subject-append \"r%zu\") payload))\n",
                             i);
            break;
        case BENCH_GATE:
            ok = append_line(&buf,
                             "(on \"sensors.*\" (-> payload-float (squelch) (publish! (subject-append \"r%zu\"))))\n",
                             i);
            break;
        case BENCH_WINDOW:
            ok = append_line(&buf,
                             "(on \"sensors.*\" (-> payload-float (moving-avg 10) (publish! (subject-append \"r%zu\"))))\n",
                             i);
            break;
        case BENCH_JSON:
            ok = mb_buf_append(&buf,
                               "(on \"sensors.*\" (json-demux! \"a\" \"b\" \"c\" payload))\n",
                               strlen("(on \"sensors.*\" (json-demux! \"a\" \"b\" \"c\" payload))\n"));
            break;
        case BENCH_MIXED:
            break;
        }
        if (!ok) {
            mb_buf_free(&buf);
            return false;
        }
    }
    if (!mb_buf_append_byte(&buf, '\0')) {
        mb_buf_free(&buf);
        return false;
    }
    *out = (char *)buf.ptr;
    *out_len = buf.len - 1;
    return true;
}

static bool subject_matches(pb_slice filter, pb_slice subject) {
    return mb_router_subject_matches((mb_slice){.ptr = (const uint8_t *)filter.ptr, .len = filter.len},
                                     (mb_slice){.ptr = (const uint8_t *)subject.ptr, .len = subject.len});
}

static bool run_program(pb_program *program, pb_slice subject, pb_slice payload,
                        uint64_t now_ms, noop_publisher *pub) {
    bool ok = true;
    for (size_t i = 0; i < program->len; i += 1) {
        pb_rule *rule = &program->rules[i];
        if (!subject_matches(rule->filter, subject)) {
            continue;
        }
        pb_eval_ctx ctx = {
            .arena = &program->scratch,
            .state = &rule->state,
            .rule_id = i,
            .now_ms = now_ms,
            .wall_ms = 0,
            .subject = subject,
            .payload = payload,
            .publish = publish_noop,
            .publish_ctx = pub,
        };
        const pb_eval_result r = pb_eval(&ctx, rule->body);
        if (r.err != PB_EVAL_OK) {
            fprintf(stderr, "patchbay bench: rule %zu failed: %s\n", i, pb_eval_error_name(r.err));
            ok = false;
        }
    }
    pb_arena_reset(&program->scratch);
    pb_arena_trim(&program->scratch, BENCH_SCRATCH_RETAIN_BYTES);
    return ok;
}

static pb_slice pick_payload(bench_mode mode, size_t i) {
    static const char *floats[] = {
        "1.0", "1.1", "1.2", "1.3", "1.4", "1.5", "1.6", "1.7",
        "2.0", "2.1", "2.2", "2.3", "2.4", "2.5", "2.6", "2.7",
    };
    const char *s = mode == BENCH_JSON ? "{\"a\":1,\"b\":2,\"c\":3}" : floats[i % (sizeof floats / sizeof floats[0])];
    return (pb_slice){.ptr = s, .len = strlen(s)};
}

static bool parse_size(const char *s, size_t *out) {
    char *end = NULL;
    const unsigned long long v = strtoull(s, &end, 10);
    if (end == s || *end != '\0') {
        return false;
    }
    *out = (size_t)v;
    return true;
}

int main(int argc, char **argv) {
    bench_mode mode = BENCH_PASS;
    if (argc > 1 && !parse_mode(argv[1], &mode)) {
        fprintf(stderr, "bad mode '%s' (want pass|gate|window|json|mixed)\n", argv[1]);
        return 2;
    }

    size_t n = 1;
    size_t pubs = 1000000;
    if (argc > 2 && !parse_size(argv[2], &n)) {
        fprintf(stderr, "bad N '%s'\n", argv[2]);
        return 2;
    }
    if (argc > 3 && !parse_size(argv[3], &pubs)) {
        fprintf(stderr, "bad PUBS '%s'\n", argv[3]);
        return 2;
    }
    if (mode != BENCH_MIXED && n == 0) {
        fprintf(stderr, "N must be > 0\n");
        return 2;
    }

    char *source = NULL;
    size_t source_len = 0;
    mb_buf file_source = {0};
    bool source_ok = false;
    if (mode == BENCH_MIXED) {
        source_ok = mb_read_file("patchbay.edn", &file_source);
        if (!source_ok) {
            perror("patchbay.edn");
        } else {
            source = (char *)file_source.ptr;
            source_len = file_source.len;
        }
    } else {
        source_ok = build_source(mode, n, &source, &source_len);
    }
    if (!source_ok) {
        return 1;
    }

    pb_program program = {0};
    if (!pb_program_load_source(&program, mode == BENCH_MIXED ? "patchbay.edn" : "<bench>", source, source_len)) {
        if (mode == BENCH_MIXED) {
            mb_buf_free(&file_source);
        } else {
            free(source);
        }
        return 1;
    }
    if (mode == BENCH_MIXED) {
        mb_buf_free(&file_source);
    } else {
        free(source);
    }

    noop_publisher pub = {0};
    char subject_buf[64];
    for (size_t i = 0; i < 10000; i += 1) {
        const int nsubj = snprintf(subject_buf, sizeof subject_buf, "sensors.k%zu", i % 1024);
        if (nsubj < 0 || (size_t)nsubj >= sizeof subject_buf) {
            pb_program_free(&program);
            return 1;
        }
        const pb_slice subject = {.ptr = subject_buf, .len = (size_t)nsubj};
        if (!run_program(&program, subject, pick_payload(mode, i), i, &pub)) {
            pb_program_free(&program);
            return 1;
        }
    }
    const uint64_t warm_emits = pub.count;
    pub.count = 0;

    const uint64_t start = mono_ns();
    for (size_t i = 0; i < pubs; i += 1) {
        const int nsubj = snprintf(subject_buf, sizeof subject_buf, "sensors.k%zu", i % 1024);
        if (nsubj < 0 || (size_t)nsubj >= sizeof subject_buf) {
            pb_program_free(&program);
            return 1;
        }
        const pb_slice subject = {.ptr = subject_buf, .len = (size_t)nsubj};
        if (!run_program(&program, subject, pick_payload(mode, i), i + 10000, &pub)) {
            pb_program_free(&program);
            return 1;
        }
    }
    const uint64_t elapsed_ns = mono_ns() - start;

    const double elapsed_s = (double)elapsed_ns / 1000000000.0;
    const double rate = (double)pubs / elapsed_s;
    const double ns_per_pub = (double)elapsed_ns / (double)pubs;
    const double emits_per_pub = (double)pub.count / (double)pubs;

    printf("mode=%s rules=%zu | %zu pubs in %.3fs = %.0f pubs/s (%.1f ns/pub, %.2f emits/pub, warm_emits=%" PRIu64 ")\n",
           mode_name(mode), program.len, pubs, elapsed_s, rate, ns_per_pub, emits_per_pub, warm_emits);

    pb_program_free(&program);
    return 0;
}
